internal/dataplane/parser/parser.go

```diff
@@ -648,20 +656,44 @@ func getServiceEndpoints(
 	return targetsForEndpoints(endpoints)
 }
 
+// getIngressClassParametersOrDefault returns the parameters for the current ingress class.
+// If the cluster operators have specified a set of parameters explicitly, it returns those.
+// Otherwise, it returns a default set of parameters.
+func getIngressClassParametersOrDefault(s store.Storer) (configurationv1alpha1.IngressClassParametersSpec, error) {
+	params, err := s.GetIngressClassParametersV1Alpha1()
+	if err != nil {
+		return configurationv1alpha1.IngressClassParametersSpec{}, err
+	}
+
+	return params.Spec, nil
+}
+
 // getEndpoints returns a list of <endpoint ip>:<port> for a given service/target port combination.
+// It also checks if the service is an upstream service either by its annotations
+// of by IngressClassParameters configuration provided as a flag.
 func getEndpoints(
 	log logrus.FieldLogger,
 	s *corev1.Service,
 	port *corev1.ServicePort,
 	proto corev1.Protocol,
 	getEndpoints func(string, string) (*corev1.Endpoints, error),
+	isSvcUpstream bool,
 ) []util.Endpoint {
 	upsServers := []util.Endpoint{}
 
 	if s == nil || port == nil {
 		return upsServers
 	}
 
+	// If service is an upstream service...
+	if isSvcUpstream || annotations.HasServiceUpstreamAnnotation(s.Annotations) {
+		// ... return its address as the only endpoint.
+		return append(upsServers, util.Endpoint{
+			Address: s.Name + "." + s.Namespace + ".svc",
+			Port:    fmt.Sprintf("%v", port.Port),
```

Comment:
I think it may be _slightly_ more idiomatic to use strconv from the standard library for this purpose:

```suggestion
			Port:    strconv.Itoa(port.Port),
```